:tada:(ansible) Add roles

waelctl opened the pull request.

waelctl committed:
:construction:(ansible) Add skeleton

waelctl committed:
:sparkles:(role) Add debian common

waelctl committed:
🐛(role) fix naming

waelctl committed:
✨(scripts) Add Elastic TLS generation

waelctl committed:
✨(role) Add Elastic role

waelctl committed:
✨(role) Add Grafana role

jmaupetit commented on elastic-tls/certutil-input.yaml:
```diff
@@ -0,0 +1,11 @@
+---
+instances:
+  - name: "elk-1"
+    ip:
+      - "X.X.X.X"
+  - name: "elk-2"
+    ip:
+      - "X.X.X.X"
+  - name: "elk-3"
```
```suggestion
  - name: "es-01"
    ip:
      - "X.X.X.X"
  - name: "es-02"
    ip:
      - "X.X.X.X"
  - name: "es-03"
```

waelctl replied:
not final version, just for testing purpose, but I'll note that

jmaupetit commented on elastic-tls/generate-tls.sh:
```diff
@@ -0,0 +1,93 @@
+    echo "--------------------------------"
+    echo
+    usage
+fi
```
What about making those two shell scripts an Ansible playbook that targets localhost instead?

waelctl replied:
because it is a pre requirements for the project, and adding a playbook ansible just to run one script shell is not relevant.
You already ask me to do a shell script or makefile for this matter :)

jmaupetit replied:
Yup. But reading your code led me to this new idea! If you prefer keeping it this way, I would suggest to merge the two scripts into one single `bin/generate-es-cluster-certificates.sh` file. Random tips we use:

- add `set -eo pipefail` as the first instruction for this script
- prefer `#!/usr/bin/env bash` shebang
- use [shellcheck](https://www.shellcheck.net) to lint your script
- define default values for your variables, _e.g._ `single` or `cluster` mode, etc.

You will find good practices we try our best to respect in the following example script: https://github.com/openfun/arnold/blob/master/bin/arnold

waelctl committed:
🚨(scripts) Fix linting and merge scripts

waelctl committed:
✨(role) Add pyenv

waelctl committed:
✨(role) Add Ralph

waelctl committed:
🎨(playbook) Add Playbook

waelctl committed:
♻️(role) Refactor pyenv

waelctl committed:
📝(docs) README

waelctl committed:
💥(all) Overall refactoring and improvement

jmaupetit commented on ansible.cfg:
```diff
@@ -0,0 +1,10 @@
+[defaults]
+roles_path = roles/
+host_key_checking = False
+fact_caching = jsonfile
+fact_caching_timeout = 3600
+fact_caching_connection = /tmp/ansible-fact-cache
+retry_files_enabled = False
+
+[ssh_connection]
+pipelining = True
```
not sure all of those settings should be enforced here

jmaupetit commented on elastic-tls/generate-tls.sh:
```diff
@@ -0,0 +1,115 @@
+check_docker() {
+  if ! command -v docker &> /dev/null
+  then
+    echo "Docker could not be found"
+    echo "Docker is required to run this script, please install"
```
```suggestion
    echo "Docker is required to run this script, please install it first."
```

jmaupetit commented on elastic-tls/generate-tls.sh:
```diff
@@ -0,0 +1,115 @@
+
+check_config() {
+  if [ ! -f ./certutil-input.yaml ]; then
+    echo "certutil-input.yaml not found"
+    echo "You must create certutil-input.yaml before running this script."
```
```suggestion
    echo "You must create the certutil-input.yaml file before running this script."
```

jmaupetit commented on elastic-tls/generate-tls.sh:
```diff
@@ -0,0 +1,115 @@
+
+help() {
+   # Display Help
+   echo "Script for generating Elastic TLS files."
```
```suggestion
   echo "Script for generating Elasticsearch TLS files."
```

jmaupetit commented on elastic-tls/generate-tls.sh:
```diff
@@ -0,0 +1,115 @@
+  docker run --rm -t \
+    --mount type=bind,source="$(pwd)",target=/tmp \
+    docker.elastic.co/elasticsearch/elasticsearch:"${ELASTIC_VERSION}" \
+    bash -c " elasticsearch-certutil ca --silent --out /tmp/my-ca.p12 --pass \"\""
```
I would suggest adding one:

```suggestion
    bash -c " elasticsearch-certutil ca --silent --out /tmp/my-ca.p12 --pass \"${ES_CA_CERT_PASSWORD}\""
```

jmaupetit commented on elastic-tls/generate-tls.sh:
```diff
@@ -0,0 +1,115 @@
+    bash -c " elasticsearch-certutil ca --silent --out /tmp/my-ca.p12 --pass \"\""
+}
+
+generate_certs() {
+  if [ "$TYPE" == "single" ]; then
+    docker run --rm -t \
+        --mount type=bind,source="$(pwd)",target=/tmp \
+        docker.elastic.co/elasticsearch/elasticsearch:"$ELASTIC_VERSION" \
+        bash -c "elasticsearch-certutil cert --silent --multiple --ca /tmp/my-ca.p12 --ca-pass \"\" --out /tmp/my-keystore.p12 --pass \"\" --in /tmp/certutil-input.yaml"
+  elif [ "$TYPE" == "cluster" ]; then
+    docker run --rm -t \
+      --mount type=bind,source="$(pwd)",target=/tmp \
+      docker.elastic.co/elasticsearch/elasticsearch:"$ELASTIC_VERSION" \
+      bash -c "elasticsearch-certutil cert --silent --ca /tmp/my-ca.p12 --ca-pass \"\" --out /tmp/my-keystore.p12 --pass \"\" --in /tmp/certutil-input.yaml"
+  else
+      echo
+      echo "The type must be single or cluster."
+      echo "--------------------------------"
+      echo
+      usage
+  fi
+}
```
What about factorizing a bit to emphasize the difference between the two commands and ease its maintenance?

```suggestion
generate_certs() {
  alias docker-run-es="docker run --rm -t \
        --mount type=bind,source=${PWD},target=/tmp \
        docker.elastic.co/elasticsearch/elasticsearch:${ELASTIC_VERSION}"
  certutil_base_cmd="""elasticsearch-certutil cert \
    --silent \
    --ca /tmp/my-ca.p12 --ca-pass "${ES_CA_CERT_PASSWORD}" \
    --out /tmp/my-keystore.p12 --pass "${ES_CERT_PASSWORD}" 
    --in /tmp/certutil-input.yaml"""
  if [ "$TYPE" == "single" ]; then
    docker-run-es bash -c "${certutil_base_cmd}"
  elif [ "$TYPE" == "cluster" ]; then
    docker-run-es bash -c "${certutil_base_cmd} --multiple"
  else
      echo
      echo "The type must be single or cluster."
      echo "--------------------------------"
      echo
      usage
  fi
}
```

jmaupetit commented on inventories/fun.yaml:
```diff
@@ -0,0 +1,60 @@
+      xpack.security.authc.realms.file.file1.order: 0
+      xpack.security.authc.realms.native.native1.order: 1
+    es_heap_size: 1g
+    es_api_basic_auth_username: "elastic"
+    es_api_basic_auth_password: "changeme"
```
Should be in an Ansible Vault.

jmaupetit commented on inventories/fun.yaml:
```diff
@@ -0,0 +1,60 @@
+        database: "[metrics-]YYYY.MM.DD"
+        url: "https://{{ hostvars[groups['elastic'][0]]['ansible_host'] }}:9200"
+        basicAuth: true
+        basicAuthUser: elastic
+        basicAuthPassword: changeme
```
ditto

jmaupetit commented on inventories/fun.yaml:
```diff
@@ -0,0 +1,60 @@
+  hosts:
+    ralph_0:
+      ansible_host: 163.172.159.6
+  vars:
```
I would suggest to refactor the inventory to ease hosts groups management in different playbooks:

```yaml
# This is an example inventory, feel free to modify it given your needs an
# constraints.
#
# Note that host names should be resolvable DNS entries. If you don't have
# permissions to modify those entries, feel free to add an `ansible_host` key
# to each hosts entry, e.g.:
#
# hosts:
#   la-stack-es-master-01:
#     ansible_host: www.xxx.yyy.zzz
#
all:
  # Common vars for all hosts
  vars:
    ansible_user: root
    debian_common_admin_user_authorized_keys:
      # paste authorized ssh keys

  children:
    # Elasticsearch nodes
    elasticsearch:
      children:
        es-master-nodes:
          hosts:
            la-stack-es-master-01:
        es-data-nodes:
          hosts:
            la-stack-es-data-01:
            la-stack-es-data-02:
      vars:
        es_config:
          node.name: "{{ inventory_hostname }}"
          cluster.name: "test-cluster"
          discovery.seed_hosts: "{{ hostvars[groups['es-master-nodes'][0]] }}:9300"
          cluster.initial_master_nodes: "{{ hostvars[groups['es-master-nodes'][0]] }}"
          http.host: 0.0.0.0
          http.port: 9200
          transport.host: 0.0.0.0
          transport.port: 9300
          bootstrap.memory_lock: true
          xpack.security.authc.realms.file.file1.order: 0
          xpack.security.authc.realms.native.native1.order: 1
        es_heap_size: 1g
        es_enable_http_ssl: true
        es_enable_transport_ssl: true
        es_ssl_keystore: "./elastic-tls/keystore/{{ inventory_hostname }}/{{ inventory_hostname }}.p12"
        es_ssl_truststore: "./elastic-tls/my-ca.p12"

    # Grafana servers
    grafana:
      hosts:
        la-stack-grafana-web-00:
      vars:
        grafana_datasources:
          - name: Elastic
            type: elasticsearch
            access: proxy
            database: "[metrics-]YYYY.MM.DD"
            url: "https://{{ hostvars[groups['es-master-nodes'][0]] }}:9200"
            basicAuth: true
            jsonData:
              tlsSkipVerify: true

    # Ralph LRS servers
    ralph:
      hosts:
        la-stack-ralph-00:
```

jmaupetit commented on playbook.yml:
```diff
@@ -0,0 +1,80 @@
+---
```
I would rename this playbook `bootstrap.yml`.

jmaupetit commented on playbook.yml:
```diff
@@ -0,0 +1,80 @@
+    - role: elastic
+      vars:
+        es_config:
+          node.master: false
```
With a refactored inventory, you can avoid repetition and factor all of those task to target : `all` and then ES master, nodes... etc.

jmaupetit commented on playbook.yml:
```diff
@@ -0,0 +1,80 @@
+      bootstrap: true
+      add_remove_keys: true
+      vars:
+        ports: ["22"]
+    - role: pyenv
+      vars:
+        pyenv_owner: admin
+        pyenv_virtualenvs:
+          - venv_name: ralph_venv
+            py_version: 3.9.15
```
Declarative, host/role related configuration variables should live in the inventory: users are not supposed to modify a playbook to achieve such tasks (see `ports` or `py_version` for example).

jmaupetit commented on inventories/fun.yaml:
```diff
@@ -0,0 +1,60 @@
+all:
+  vars:
+    ansible_user: root
```
Wouldn't it be preferable?

```suggestion
    ansible_become: yes
```

jmaupetit commented on inventories/fun.yaml:
```diff
@@ -0,0 +1,60 @@
+  vars:
+    ansible_user: root
+    debian_common_admin_user_authorized_keys:
+      - ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAcLMHQbGDsGDHQ90eV7Tt1/lm98b3hgjn1cFB9xKlNU julien.maupetit@fun-mooc.fr
```
Oops :sweat_smile:

jmaupetit commented on roles/debian-common/defaults/main.yml:
```diff
@@ -0,0 +1,61 @@
+  - git
+  - htop
+  - acl
+  - dbus
```
I would only install required dependencies and allow to extend it via a configuration variable. WDYT?

```suggestion
  - bash-completion
  - apt-transport-https
  - ca-certificates
  - software-properties-common
  - curl
```

jmaupetit commented on roles/debian-common/defaults/main.yml:
```diff
@@ -0,0 +1,61 @@
+  - libsqlite3-dev
+  - libreadline-dev
+  - libffi-dev
+  - libbz2-dev
```
ditto

python-related dependencies should only be installed in python-related hosts, not all!

jmaupetit commented on roles/debian-common/defaults/main.yml:
```diff
@@ -0,0 +1,61 @@
+  - libffi-dev
+  - libbz2-dev
+
+debian_common_supplementary_packages: []
```
:+1:

jmaupetit commented on roles/debian-common/tasks/bootstrap-firewall.yml:
```diff
@@ -0,0 +1,29 @@
+    port: "{{ item }}"
+    proto: tcp
+  with_items:
+    - "{{ ports | default(['22']) }}"
```
I think this variable name should be more explicit, _e.g._ `firewall_allowed_ports`?

jmaupetit commented on roles/debian-common/tasks/bootstrap-firewall.yml:
```diff
@@ -0,0 +1,29 @@
+    src:  "{{ item.value.ip }}"
+    port: "{{ item.value.port }}"
+    proto: tcp
+  with_dict: "{{ port_ips | default({}) }}"
```
ditto for `port_ips`

jmaupetit commented on roles/elastic/templates/elasticsearch.j2:
```diff
@@ -0,0 +1,67 @@
+{% endif %}
+
+# Maximum number of VMA (Virtual Memory Areas) a process can own
+# When using Systemd, this setting is ignored and the 'vm.max_map_count'
```
Shouldn't we set this system-wide for all ES hosts?

jmaupetit commented on roles/elastic/filter_plugins/custom.py:
```diff
@@ -0,0 +1,75 @@
+import re
+import os.path
+from six import string_types
+
```
The python bytecode (*.pyc) file should be git-ignored (so does the `__pycache__` directory).

jmaupetit reviewed (commented):
This is a great start!

Global remarks:

- host-specific vars should be defined in the inventory first, not in a playbook or tasks list
- secret variables should be stored in an Ansible vault for each service
- this project should be generic and configurable, one should avoid hard-coding things in the ansible configuration or inventory

Thank you!

waelctl committed:
🧑‍💻(all) overall improvements

waelctl committed:
👔(all) Improve business logic

jmaupetit commented on roles/ralph/README.md:
```diff
@@ -0,0 +1,27 @@
+      roles:
+         - role: ralph
+           become_user: admin
+           become: true
```
Is this RestructuredText? :sweat_smile:

waelctl committed:
🍻(all) Refactor elastic and playbook

jmaupetit commented on requirements.txt:
```diff
@@ -0,0 +1 @@
+jmespath==1.0.1
```
Ansible is missing :wink:

jmaupetit replied:
I've added finally added it.

jmaupetit committed:
Add missing Ansible dependency

jmaupetit committed:
Ignore python virtual environments

jmaupetit committed:
Clean generate-tls code and usage

jmaupetit committed:
Who's still running python 2?

jmaupetit committed:
Fix grafana configuration

jmaupetit committed:
draft placeholders to run ralph web server

jmaupetit reviewed (commented):
I've added to commit for minor fixes / improvements.

The Elasticsearch cluster and Grafana application deployments are validated :+1: 

One last thing to do is to:

- add a systemd service that runs `gunicorn` (I've added the command line to run in the `systemd.yml` task)
- add a role handler so that we can restart the service when upgrading Ralph, a dependency, or a configuration during deployment.

waelctl committed:
✨ (ansible) Add systemd for ralph

Add Vagrantfile for local testing the LA Ansible stack
Add systemd service for Ralph
Refactor the inventories

waelctl committed:
✨(role) add Superset and pgsql
